File: squareplay/src/bezier.py

import math
from mathutils import Matrix, Vector

class Bezier:

  def __init__(self,x1,y1,ctrlx1,ctrly1,ctrlx2,ctrly2,x2,y2):
    self.x1 = x1
    self.y1 = y1
    self.ctrlx1 = ctrlx1
    self.ctrly1 = ctrly1
    self.ctrlx2 = ctrlx2
    self.ctrly2 = ctrly2
    self.x2 = x2
    self.y2 = y2
    self.calculatecoefficients()

  def calculatecoefficients(self):
    self.cx = 3.0*(self.ctrlx1-self.x1)
    self.bx = 3.0*(self.ctrlx2-self.ctrlx1) - self.cx
    self.ax = self.x2 - self.x1 - self.cx - self.bx

    self.cy = 3.0*(self.ctrly1-self.y1)
    self.by = 3.0*(self.ctrly2-self.ctrly1) - self.cy
    self.ay = self.y2 - self.y1 - self.cy - self.by

  #  Return the location along the curve given "t" between 0 and 1
  def translation(self,t):
    x = self.x1 + t*(self.cx + t*(self.bx + t*self.ax))
    y = self.y1 + t*(self.cy + t*(self.by + t*self.ay))
    return Vector([x,y,0])
  def translationMatrix(self,t):
    return Matrix.Translation(self.translation(t))

  #  Return the direction. which is the derivative, given "t" between 0 and 1
  def direction(self,t):
    x = self.cx + t*(2.0*self.bx + t*3.0*self.ax)
    y = self.cy + t*(2.0*self.by + t*3.0*self.ay)
    return Vector([x,y,0])
  def directionMatrix(self,t):
    return Matrix.Translation(self.direction(t))

  #  Return the angle of the derivative given "t" between 0 and 1
  #  This is the angle that the curve is moving to
  def rotation(self,t):
    x = self.cx + t*(2.0*self.bx + t*3.0*self.ax)
    y = self.cy + t*(2.0*self.by + t*3.0*self.ay)
    return math.atan2(y,x)
  def rotationMatrix(self,t):
    return Matrix.Rotation(self.rotation(t),4,'Z')

  #  Return the angle of the second derivative given "t"
  #  For determining the turning direction
  def turn(self,t):
    x = 2.0*self.bx + t*6.0*self.ax
    y = 2.0*self.by + t*6.0*self.ay
    return math.atan2(y,x)
  def turnMatrix(self,t):
    return Matrix.Rotation(self.turn(t),4,'Z')

File: src/bezier.js
/*

    Copyright 2017 Brad Christie

    This file is part of Taminations.

    Taminations is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published
    by the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Taminations is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with Taminations.  If not, see <http://www.gnu.org/licenses/>.

 */

define(['affinetransform'],function(AffineTransform) {

  //  Bezier class
  Bezier = function (x1,y1,ctrlx1,ctrly1,ctrlx2,ctrly2,x2,y2)
  {
    this.x1 = x1;
    this.y1 = y1;
    this.ctrlx1 = ctrlx1;
    this.ctrly1 = ctrly1;
    this.ctrlx2 = ctrlx2;
    this.ctrly2 = ctrly2;
    this.x2 = x2;
    this.y2 = y2;
    this.calculatecoefficients();
  };

  Bezier.prototype.calculatecoefficients = function()
  {
    this.cx = 3.0*(this.ctrlx1-this.x1);
    this.bx = 3.0*(this.ctrlx2-this.ctrlx1) - this.cx;
    this.ax = this.x2 - this.x1 - this.cx - this.bx;

    this.cy = 3.0*(this.ctrly1-this.y1);
    this.by = 3.0*(this.ctrly2-this.ctrly1) - this.cy;
    this.ay = this.y2 - this.y1 - this.cy - this.by;
  };

  //  Compute X, Y values for a specific t value
  Bezier.prototype.xt = function(t) {
    return this.x1 + t*(this.cx + t*(this.bx + t*this.ax))
  }
  Bezier.prototype.yt = function(t) {
    return this.y1 + t*(this.cy + t*(this.by + t*this.ay))
  }
  //  Compute dx, dy values for a specific t value
  Bezier.prototype.dxt = function(t) {
    return this.cx + t*(2.0*this.bx + t*3.0*this.ax)
  }
  Bezier.prototype.dyt = function(t) {
    return this.cy + t*(2.0*this.by + t*3.0*this.ay)
  }
  Bezier.prototype.angle = function(t) {
    return Math.atan2(this.dyt(t),this.dxt(t))
  }

  //  Return the movement along the curve given "t" between 0 and 1
  Bezier.prototype.translate = function(t)
  {
    var x = this.xt(t)
    var y = this.yt(t)
    return AffineTransform.getTranslateInstance(x,y);
  };

  //  Return the angle of the derivative given "t" between 0 and 1
  Bezier.prototype.rotate = function(t)
  {
    var theta = this.angle(t)
    return AffineTransform.getRotateInstance(theta);
  };

  //  Return turn direction at end of curve
  Bezier.prototype.rolling = function()
  {
    //  Check angle at end
    var theta = this.angle(1.0)
    //  If it's 180 then use angle at halfway point
    if (Math.anglesEqual(theta,Math.PI))
      theta = this.angle(0.5)
    //  If angle is 0 then no turn
    if (Math.anglesEqual(theta,0))
      return 0
    else
      return theta
  }

  //  Return the angle of the 2nd derivative given "t" between 0 and 1
  Bezier.prototype.turn = function(t)
  {
    var x = 2.0*this.bx + t*6.0*this.ax;
    var y = 2.0*this.by + t*6.0*this.ay;
    var theta = Math.atan2(y,x);
    return theta;
  };


  Bezier.prototype.toString = function()
  {
    return '[ '+this.x1.toFixed(1)+' '+this.y1.toFixed(1)+' '+
    this.ctrlx1.toFixed(1)+' '+this.ctrly1.toFixed(1)+' '+
    this.ctrlx2.toFixed(1)+' '+this.ctrly2.toFixed(1)+' '+
    this.x2.toFixed(1)+' '+this.y2.toFixed(1)+' ]';
  };

  return Bezier;

});
